A better story for canary jobs (yellow CI)

jonhoo opened the pull request:
This PR does two things.

First, it adds a new smoke test to the mix: `cargo check -- -D warnings`, which will error on _any_ warning from `rustc`.

Second, it makes it easy to opt out of all canaries (basically anything that is `allow_fail`) by setting 
```yaml
parameters:
  canaries: false
```
This will currently disable:
 - rustfmt on beta
 - clippy on beta
 - the new `-D warnings` test
 - test on nightly

This parameter is intended for users that worry about long-running or highly parallel CI, yellow CI ("partially successful").

jonhoo committed:
Add warn-free clippy smoketest

jonhoo committed:
Make it easy to disable all smoke tests

epage commented on azure/style.yml:
```diff
@@ -1,22 +1,31 @@
 parameters:
   setup: []
+  smoketests: true
```
What is the intention of `smoketests`?

It doesn't look like the meaning I'm used to [which is to determine whether further testing should be done](https://en.wikipedia.org/wiki/Smoke_testing_(software)).  `cargo check` seems like it acts like a smoke tests.

jonhoo replied:
Ah, sorry, you're right, smoke test isn't really the right word for this.

The meaning I was going for was something that, when it fails, won't prevent the build, but _will_ show up as a warning for the build (a yellow "CI partially succeeded" badge in Azure Pipelines terms). What do you think a better word for this is?

epage replied:
Naming is hard.

You've used `canary` in other places and kind of works but by itself doesn't convey enough.

epage commented on azure/cargo-clippy.yml:
```diff
@@ -2,13 +2,20 @@ parameters:
   rust: stable
   allow_fail: false
   setup: []
+  fail_all: false
```
What is the intent behind `fail_all`?  It seems to add warnings but the parameter name doesn't imply that to me.

jonhoo replied:
`-D warnings` turns all warnings into hard errors. `fail_all` was intended to indicate that "this will fail on any and all issues", but maybe there's a better wording?

epage replied:
`include_warnings`?

jonhoo replied:
Hmm.. It's not that it _includes_ warnings, but rather that it treats all warnings as errors. `warnings_as_errors` is pretty verbose though, so I'd prefer to find something shorter.

epage replied:
`include_warnings` as in "check clippy and warnings" though I also see how it doesn't clarify they are errors.  Unsure if its needed?

epage commented on azure/style.yml:
```diff
@@ -1,22 +1,31 @@
+    - template: cargo-clippy.yml
+      parameters:
+        name: clippy_nowarnings
+        fail_all: true
```
This is confusing.  `clippy_nowarnings` but I thought `fail_all` implies warnings?

jonhoo replied:
`nowarnings` here is to imply that "no warnings are allowed". I agree this is misleading.

epage replied:
I love how we keep reading the opposite thing in each others naming :)

epage commented on azure/style.yml:
```diff
@@ -1,22 +1,31 @@
-      setup: ${{ parameters.setup }}
-      rust: beta
-      allow_fail: true
+  - ${{ if eq(parameters.smoketests, 'true') }}:
```
imo warnigns should not be hidden behind `smoketests`.  if its about build stability, clippy and warnings are probably about the same level of instability and just as important as each other

jonhoo replied:
So, this parameter was more intended to be "don't run extra things/things whose failures are ignored". A failed `cargo clippy` _should_ fail the build (imo), but a warning should not (in part because they can be caused by semver-compatbile changes to dependencies, and thus are out of your control).

jonhoo committed:
Use cargo check with -Dwarnings over clippy

Enabling _all_ clippy warnings is too hardcore

jonhoo committed:
Canaries is better than smoketests

They're not really smoke tests since they don't pre-empt other tests
from running. Also, canaries are yellow, just like your CI will be.

This also turns cargo +beta test into an allow_fail, because it is also
a canary.

jonhoo committed:
Slightly more concise name for -Dwarnings

jonhoo committed:
cargo check needs setup

epage commented on azure/cargo-check.yml:
```diff
@@ -8,12 +8,12 @@ jobs:
   ${{ if ne(parameters.rust, 'stable') }}:
     ${{ if eq(parameters.err_on_warn, 'true') }}:
-      displayName: "env RUSTFLAGS=-Dwarnings cargo +${{ parameters.rust }} check"
+      displayName: "cargo +${{ parameters.rust }} check -Dwarnings"
```
Should we allow `deprecated`?  It'll make things more stable.

jonhoo replied:
This is already `allow_fail`, so I'd be inclined to still give deprecation warnings. They're things you should probably fix, just like other warnings.

epage replied:
1. This template does not have an `allow_fail`
2. This assumes that the caller will always pass `allow_fail` with `err_on_warn`.  In my use case, I do not want to be doing that.

jonhoo replied:
Ah, good catch! Fixed 1 in latest commit.

As for 2, it doesn't really assume that. If you specify `err_on_warn` without `allow_fail`, you are specifically saying "I want to treat warnings as CI errors". Seems reasonable for that to be a valid option?

epage replied:
While so far I've gotten away with `deprecated` being enabled, you've brought up the concern over dependencies breaking people.  If we want to take that seriously and support people who want a pinned warnings check, then we'd need to disable that warning.

jonhoo replied:
My thinking here was that either you care about warnings, or you do not (this is `err_on_warn`). If you do, you either want them to cause red CI or you want them to cause yellow CI (this is `allow_fail`). We could add a third branch point which is "care about all but warnings caused by changes in dependencies" I suppose (what would it be called)? My guess is that that option is less common, but I have limited data on that.

epage replied:
My care abouts
- Warnings make CI red
- Transient state (like dependencies deprecating things when you don't have a Cargo.lock checked in) should not make my CI red
  - Personally, I've never had a dependency make something deprecated so I've not worried about that state before.  This is more in theoretically possible land
- I should be able to maintain my MSRV ie deprecation warnings for Rust can only be emitted by the MSRV or else I can't get a "non-red" to approve my submissions.

As long as those conditions are met, I'm fine.

jonhoo replied:
Okay, so, I _think_ this satisfies all of those. Specifically, the default `stages.yml` will never make your CI red from warnings, it will only ever make it yellow. You _can_ choose to directly depend on `cargo-check.yml`, and then you can choose whether you want warnings to cause red CI (`err_on_warn: true` + `allow_fail: false`), yellow CI (`err_on_warn: true` + `allow_fail: true`), or green CI (`err_on_warn: false` + `allow_fail: false`; the default).

Thus, anything that'd make your CI red would be by choice in setting up your pipeline.

Now, what this PR does _not_ give you the ability to do is to run CI and have it _fail_ on warnings that are _not_ due to dependencies. I think that's specifically something you would like, but I also think that's something that's better handled in a new PR where we can explicitly figure out which warnings that would require allowing.

So, with all that said, my inclination would be to merge and then open an issue tracking "red CI from non-dependency warnings".

jonhoo committed:
Support allow_fail on cargo check

jonhoo committed:
Remove remains of reverted change

djc commented:
FWIW, I like the idea of (optionally) erroring stable CI for warnings. I think denying warnings at compile time is too annoying for my development experience, but I want code to be warning-free before it hits master.

jonhoo commented:
@djc there are two tricky parts about this:

 - Commits that did not emit warnings in the past may _start_ emitting warnings in future versions of `stable`, which would cause a CI failure in, say, an unrelated PR.
 - Semver-compatible changes in _dependencies_ (macro changes and deprecations in particular) can cause a previously-fine commit to suddenly stop passing if you don't check in `Cargo.lock`.

These taken together mean that failing CI on warnings, even on `stable` is risky business. The open discussion above is basically about what policy to take on this. It might be possible to have a "error on warnings with a pinned Rust version + allow particular warnings related to dependencies", but it's finicky and somewhat brittle.

jonhoo committed:
Merge branch 'master' into smoketest

jonhoo committed:
Avoid a repeated if
